File: 프로그래머스/1/42840. 모의고사/모의고사.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
/* 
아래 패턴 반복해서 문제 찍음
1, 2, 3, 4, 5
2, 1, 2, 3, 2, 4, 2, 5
3, 3, 1, 1, 2, 2, 4, 4, 5, 5
=> 가장 많은 문제를 맞힌 사람? 
*/

// 정답 배열
vector<int> solution(vector<int> answers) { // 최대 10,000문제
    vector<int> answer; // 가장 많은 문제를 맞힌 사람
    
    // 찍는 패턴
    vector<int> first = {1, 2, 3, 4, 5};
    vector<int> second = {2, 1, 2, 3, 2, 4, 2, 5}; 
    vector<int> third = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};
    
    vector<pair<int, int>> corr(3);// 맞은 개수
    for(int i = 0; i < 3; i++){
        corr[i].second = i+1; corr[i].first = 0;
    }
    
    for(int i = 0; i < answers.size(); i++){
        if(answers[i] == first[i%first.size()]) corr[0].first++;
    }
    
    for(int i = 0; i < answers.size(); i++){
        if(answers[i] == second[i%second.size()]) corr[1].first++;
    }
    
    for(int i = 0; i < answers.size(); i++){
        if(answers[i] == third[i%third.size()]) corr[2].first++;
    }
    
    sort(corr.begin(), corr.end(), greater<pair<int, int>>()); 
    
    int best = corr[0].first;
    answer.push_back(corr[0].second); 
    for(int i = 1; i <= 2; i++){
        if(corr[i].first == best) answer.push_back(corr[i].second);
    }
    
    sort(answer.begin(), answer.end()); // 오름차순 정렬
    return answer;
}

File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.*; 
/*
정답 배열이 주어졌을때, 가장 많은 문제를 맞힌 사람은? 
가장 높은 점수 받은 사람이 여럿일 경우, return 하는 값을 오름차순으로!!!
*/

class Solution {
    public int[] solution(int[] answers) {
        int[] answer = {};
        
        int[] a = {1, 2, 3, 4, 5}; int a_idx = 0; int aa = 0;
        int[] b = {2, 1, 2, 3, 2, 4, 2, 5}; int b_idx = 0; int bb = 0;
        int[] c = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5}; int c_idx = 0; int cc = 0;
        // 하나하나 보기
        for(int i = 0; i < answers.length; i++){
            int ans = answers[i];
            if (answers[i] == a[i % a.length]) aa++;
            if (answers[i] == b[i % b.length]) bb++;
            if (answers[i] == c[i % c.length]) cc++;
        }
        
        List<List<Integer>> res = new ArrayList<>(); 
        res.add(Arrays.asList(aa, 1));
        res.add(Arrays.asList(bb, 2));
        res.add(Arrays.asList(cc, 3));
        // 내림차순 정렬
        Collections.sort(res, (s, k) -> Integer.compare(k.get(0), s.get(0)));
        
        List<Integer> ans = new ArrayList<>(); 
        
        int max = res.get(0).get(0);
        for(int i = 0; i < 3; i++){
            if(res.get(i).get(0) == max) ans.add(res.get(i).get(1));
            else break;
        }
        
        return answer = ans.stream().mapToInt(i -> i).toArray();
    }
}